samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -230,22 +242,39 @@ public String getNextOpId(OpCode opCode) {
    * @param <M> the type of messages in the intermediate {@link MessageStream}
    * @return  the intermediate {@link MessageStreamImpl}
    */
-  @VisibleForTesting
-  <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
+  public <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
     Preconditions.checkState(!inputOperators.containsKey(streamId) && !outputStreams.containsKey(streamId),
         "getIntermediateStream must not be called multiple times with the same streamId: " + streamId);
 
-    if (serde == null) {
-      LOGGER.info("Using default serde for intermediate stream: " + streamId);
-      serde = (Serde<M>) defaultSerde;
+    Serde streamSerde;
+    if (serde != null) {
+      streamSerde = serde;
+    } else {
+      LOGGER.warn("No serde defined for intermediate stream: " + streamId +
+          ". Key and message serdes configured for the job.default.system will be used.");
+      streamSerde = null;
     }
 
     if (isBroadcast) broadcastStreams.add(streamId);
-    boolean isKeyed = serde instanceof KVSerde;
-    KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+
+    boolean isKeyed;
+    KV<Serde, Serde> kvSerdes;
+    if (streamSerde == null) { // if no explicit or default system descriptor serde available
+      isKeyed = true; // assume keyed stream
+      kvSerdes = new KV<>(null, null); // and that key and msg serdes are provided for job.default.system in configs
+    } else {
+      isKeyed = streamSerde instanceof KVSerde;
+      kvSerdes = getKVSerdes(streamId, streamSerde);
+    }
+
+    InputTransformer<?> transformer = null;
+    if (defaultSystemDescriptor instanceof TransformingSystemDescriptor) {
```

Comment:
Should there have been a check for ExpandingSystemDescriptor here? If so, then I think this would be the evidence for the above "instanceof" comment in InputDescriptor.